### PR TITLE
Polkadot: Make the current inflation formula adjustable

### PR DESCRIPTION
**Pull request by kianenigma**
Counterpart of https://github.com/polkadot-fellows/runtimes/pull/364.

Similarly, it has no impact on the inflation as-is, and just exposes the params to the Root track. 

## UI Consideration

Similarly, it adds `Inflation_experimental_inflation_info` Runtime API, which can be used by UIs to fetch information about the inflation, without needing to rely on storage items.

## Example

- With current parameters: 

```
Inflation ==> staking = 402244.054 DOT (4,022,440,540,915,948) / leftover = 4178.285 DOT (41,782,856,835,356)
```

- With `UseLegacyAuctions = false`, which means we no longer set aside around 15% of tokens for auctions. This makes the difference between actual staked and ideal staking rate (staking inefficiency) grow, and therefore reduce the staking reward and increase the treasury income. 

```
Inflation ==> staking = 340780.189 DOT (3,407,801,892,871,273) / leftover = 65642.150 DOT (656,421,504,843,446)
```

### CODE REVIEW

File: relay/kusama/src/lib.rs [hunk 1]
```diff
@@ -1919,7 +1922,7 @@ impl Runtime {
 		};
 
 		// We assume un-delayed 6h eras.
-		let era_duration = 6 * HOURS;
+		let era_duration = 6 * (HOURS as Moment) * MILLISECS_PER_BLOCK;
```
Comment (kianenigma) [inline relay/kusama/src/lib.rs:1925]: unfortunately there was a small mistake here as well, but since it is read-only and experimental, it is fine.

File: relay/polkadot/Cargo.toml [hunk 2]
```diff
@@ -184,9 +190,11 @@ std = [
 	"pallet-xcm-benchmarks?/std",
 	"pallet-xcm/std",
 	"polkadot-parachain-primitives/std",
+	"polkadot-parachain-primitives/std",
```
Comment (ggwpez) [inline relay/polkadot/Cargo.toml:193]:
```suggestion
```
Hm Zepter is not detecting this?

File: relay/polkadot/src/lib.rs [hunk 3]
```diff
@@ -619,20 +619,121 @@ impl pallet_bags_list::Config<VoterBagsListInstance> for Runtime {
...
+
+		/// Falloff used to calculate era payouts.
+		#[codec(index = 3)]
+		pub static Falloff: Perquintill = Perquintill::from_percent(5);
+
+		/// Whether to use auction slots or not in the calculation of era payouts, then we subtract
+		/// `num_auctioned_slots.min(60) / 300` from `ideal_stake`.
+		///
+		/// That is, we assume up to 60 parachains that are leased can reduce the ideal stake by a
+		/// maximum of 30%.
```
Comment (gpestana) [inline relay/polkadot/src/lib.rs:652]:
```suggestion
		/// maximum of 20%.
```

As `60/300 = .2`

Comment (gpestana) [reply relay/polkadot/src/lib.rs:652]: And the removed code expects 20%, so it's a comment typo not a parameter mismatch.

File: relay/polkadot/src/lib.rs [hunk 4]
```diff
@@ -619,20 +619,121 @@ impl pallet_bags_list::Config<VoterBagsListInstance> for Runtime {
...
+	}
+}
+
+impl pallet_parameters::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	type RuntimeParameters = RuntimeParameters;
+	type AdminOrigin = DynamicParameterOrigin;
+	type WeightInfo = weights::pallet_parameters::WeightInfo<Runtime>;
+}
+
```
Comment (gpestana) [inline relay/polkadot/src/lib.rs:703]:
```suggestion
/// Defines how much should the inflation be for an era given its duration.
```

**Review by gpestana — APPROVED** [review]

File: relay/polkadot/src/lib.rs [hunk 5]
```diff
@@ -2107,7 +2132,48 @@ mod benches {
...
+			.unwrap_or((0, 0));
+		let stake_able_issuance = Balances::total_issuance();
+
+		let ideal_staking_rate = dynamic_params::inflation::IdealStake::get();
+		let inflation = if dynamic_params::inflation::UseAuctionSlots::get() {
+			let auctioned_slots = parachains_paras::Parachains::<Runtime>::get()
+				.into_iter()
+				// all active para-ids that do not belong to a system chain is the number of
+				// parachains that we should take into account for inflation.
+				.filter(|i| *i >= 2000.into())
```
Comment (Ank4n) [inline relay/polkadot/src/lib.rs:2150]: replace 2000 with `polkadot_primitives::LOWEST_PUBLIC_ID`?

**Review by Ank4n — APPROVED** [review]

**Review by ggwpez — APPROVED** [review]

File: relay/kusama/src/lib.rs [hunk 6]
```diff
@@ -1919,11 +1922,11 @@ impl Runtime {
 		};
 
 		// We assume un-delayed 6h eras.
-		let era_duration = 6 * HOURS;
+		let era_duration = 6 * 60 * 60 * 1000;
```
Comment (bkchr) [inline relay/kusama/src/lib.rs:1925]: Why was this changed? And is the comment now incorrect?

**Review by bkchr — APPROVED** [review]

File: relay/kusama/src/lib.rs (hunk 6 above)
Comment (kianenigma) [reply relay/kusama/src/lib.rs]: unified them in a better way in [a5d3ed3](https://github.com/polkadot-fellows/runtimes/pull/443/commits/a5d3ed3cbd0ab5a5f93e44242805ecc4e69b3dd2)

**Comment by ggwpez** [comment]
@Bullrich  ^

**Comment by Bullrich** [comment]
@ggwpez it seems it was a one time problem, as the method worked with my comment:
image
